<a href="https://colab.research.google.com/github/Argentan/DMA_LAB2/blob/master/tutoriales/02_intro_python2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Importación de librerías

Importar una librería entera


In [1]:
import random

Puedo asignar un nombre corto para facilitar la llamada de funciones

In [2]:
import pandas as pd

Importar solo una función de una librería

In [3]:
from random import sample, shuffle

Si no tengo instalada una librería puedo hacerlo de varias maneras. La más simple usando **pip**.

El "**!**" lo usamos en Jupyter Notebooks para indicarle que es un comando de consola.

In [4]:
#!pip install seaborn
import seaborn

Otra alternativa es con **apt-get**, por ejemplo, para esta librería que nos ayuda a leer archivos de compresión 7zip

In [5]:
# https://pypi.python.org/pypi/libarchive
# !apt-get -qq install -y libarchive-dev && pip install -U libarchive

También puedo ver la versión de la Librería que tengo usando **pip show**

In [6]:
!pip show tensorflow

Name: tensorflow
Version: 2.11.0
Summary: TensorFlow is an open source machine learning framework for everyone.
Home-page: https://www.tensorflow.org/
Author: Google Inc.
Author-email: packages@tensorflow.org
License: Apache 2.0
Location: /usr/local/lib/python3.9/dist-packages
Requires: absl-py, astunparse, flatbuffers, gast, google-pasta, grpcio, h5py, keras, libclang, numpy, opt-einsum, packaging, protobuf, setuptools, six, tensorboard, tensorflow-estimator, tensorflow-io-gcs-filesystem, termcolor, typing-extensions, wrapt
Required-by: 


También se puede indicar una versión específica de la librería

In [7]:
#!pip install tensorflow==1.2

# Estructuras de control

## Comparadores lógicos

| Operador | Acción |
| -------- | ------ |
| ==       | igual  |
| > =       | mayor o igual |
| >        | mayor estricto |
| < =      | menor o igual |
| <     | menor estricto |
| ! = | desigual|
| or | o |
| and | y |


In [8]:
1 == 1

True

In [9]:
1 == 2

False

In [10]:
1 == 1.0

True

In [11]:
"hola" == "hola"

True

In [12]:
"hola" == "Hola"

False

In [13]:
"hola" != "Hola"

True

In [14]:
(1 == 1) and (2 == 2)

True

In [15]:
(1 == 1) and (2 != 2)

False

In [16]:
(1 == 1) or (2 != 2)

True

In [17]:
(1 != 1) or (2 != 2)

False

In [18]:
not (     (1 != 1) or (2 != 2)       )

True

In [19]:
not (1 == 1) or (2 == 2)

True

### El operador "not in"

In [20]:
'Juan' in ['Pedro', 'Lucia', 'Juan', 'Maria']

True

In [21]:
'Juan' not in ['Pedro', 'Lucia', 'Juan', 'Maria']

False

In [22]:
'Julieta' not in ['Pedro', 'Lucia', 'Juan', 'Maria']

True

### Asignación con operador aritmético

| Operador | Equivalente |
| -------- | ----------- |
| `a += 1` | `a = a + 1` |
| `a -= 1` | `a = a - 1` |
| `a *= 1` | `a = a * 1` |
| `a /= 1` | `a = a / 1` |
| `a %= 1` | `a = a % 1` |

In [23]:
a = 'Hello'
a += ' world!'
a

'Hello world!'

In [24]:
a = 2
print(a)

a *= 2
print(a)

a *= 2
print(a)

2
4
8


## Condicionales

In [25]:
persona = "Juan"
if (persona == "Juan"):
    print("Juan encontrado")

Juan encontrado


In [26]:
if (persona == "Pedro"):
    print("Pedro encontrado")
else:
    print("Pedro no encontrado")

Pedro no encontrado


In [27]:
if (persona == "Pedro"):
    print("encontrado Pedro")
elif (persona == "Juan"): 
    print("encontrado Juan")
else:
    print("no encontrado")

encontrado Juan


## Función `range`

`range(stop) -> range object`

`range(start, stop[, step]) -> range object`

Devuelve un objeto que produce una secuencia de enteros desde start(inclusive)
hasta stop (exclusivo) de acuerdo a step.

`range(i, j) produce i, i+1, i+2, ..., j-1`

start por defecto es 0 y se omite stop!

`range(4) produce 0, 1, 2, 3`

Cuando se especifica `step`, especifica el incremento (o decremento).

In [28]:
rango = range(5)    # stop=5
type(rango)

range

In [29]:
range(0, 5)      # start = 0 ,  stop = 5

range(0, 5)

In [30]:
# Crear una lista de números impares del 1 al 10
lista = list(range(1, 11, 2))
lista

[1, 3, 5, 7, 9]

## Loops

### Loops usando for

In [31]:
for a in [1, 2, 3]:
    print(a)

1
2
3


In [32]:
lista = [1, 2, 3]
for a in lista:
    print(a)

1
2
3


In [33]:
diccionario = {"k1": 1, "k2": 2, "k3": 3}
for a in diccionario:
    print(a)

k1
k2
k3


In [34]:
for a in diccionario.keys():
    print(a)

k1
k2
k3


In [35]:
for a in diccionario.values():
    print(a)

1
2
3


In [36]:
diccionario = {"k1": 1, "k2": 2, "k3": 3}
for a in diccionario:
    print(a, diccionario[a])

k1 1
k2 2
k3 3


In [37]:
diccionario.items()

dict_items([('k1', 1), ('k2', 2), ('k3', 3)])

In [38]:
diccionario = {"k1": 1, "k2": 2, "k3": 3}
for k, v in diccionario.items():
    #print(f"clave: {k}, valor: {v}")
    #print("clave:", k, "valor:", v)
    print (f"esto es una cadena y el valor de k es igual a {k}")

esto es una cadena y el valor de k es igual a k1
esto es una cadena y el valor de k es igual a k2
esto es una cadena y el valor de k es igual a k3


In [39]:
v1, v2 = 1, 2
print(v1)
print(v2)

1
2


In [40]:
len(lista)

3

In [41]:
range(len(lista))

range(0, 3)

In [42]:
lista = ["a", "b", "c"]
for i in range(len(lista)):
    print(f"el indice {i} tienen un valor {lista[i]}")

el indice 0 tienen un valor a
el indice 1 tienen un valor b
el indice 2 tienen un valor c


In [43]:
lista

['a', 'b', 'c']

In [44]:
for indice, elemento in enumerate(lista):
    print(f"el indice {indice} tiene un valor {elemento}")

el indice 0 tiene un valor a
el indice 1 tiene un valor b
el indice 2 tiene un valor c


### Loops usando While 

In [45]:
from random import random

In [46]:
cent = 0
print(cent)
while cent < 0.7:
    cent = random()
    print("nuevo cent:", cent)    

0
nuevo cent: 0.1344670138953813
nuevo cent: 0.9119344869000302


### Loop de múltiples listas usando zip()


In [48]:
name = ['Pedro', 'Lucia', 'Juan', 'María', 'José']
age = [8, 9, 3, 5]

for n, a in zip(name, age):
  print(f'La edad de {n} es {a}')

La edad de Pedro es 8
La edad de Lucia es 9
La edad de Juan es 3
La edad de María es 5


In [49]:
list(zip('abcdefg', range(10), range(10)))

[('a', 0, 0),
 ('b', 1, 1),
 ('c', 2, 2),
 ('d', 3, 3),
 ('e', 4, 4),
 ('f', 5, 5),
 ('g', 6, 6)]

La función zip() en Python toma como entrada una o más secuencias (por ejemplo, listas o tuplas) y devuelve una secuencia de tuplas donde cada tupla contiene un elemento de cada una de las secuencias originales.

En el ejemplo list(zip('abcdefg', range(10), range(10))), se están usando tres secuencias como argumentos de zip(). La primera secuencia es la cadena de caracteres 'abcdefg', la segunda y tercera secuencias son dos rangos (range) de números enteros que van del 0 al 9.

La función zip() crea una secuencia de tuplas que combina elementos de las tres secuencias de entrada.

La función list() convierte esta secuencia de tuplas en una lista de tuplas, y se imprime usando la instrucción print().

### List Comprehension

Una lista por comprensión es una construcción sintáctica disponible en algunos lenguajes de programación para crear una lista basada en listas existentes. Sigue la forma de la notación matemática de creación de conjuntos (comprensión de conjuntos), por ejemplo:

$ S = \{2 x | x \in \mathbb{N}, x^2 > 3 \} $

En teoría de conjuntos, la notación anterior se lee como "$S$ es el conjunto de todos los números $2x$, tal que $x$ es un número natural y $x^2$ es mayor que 3".

### Sintaxis

newlist = [expression for item in iterable if condition == True]

In [50]:
fruits = ["apple", "banana", "cherry", "kiwi", "mango"]
fruits

['apple', 'banana', 'cherry', 'kiwi', 'mango']

In [51]:
newlist = [x for x in fruits if "a" in x]

print(newlist)

['apple', 'banana', 'mango']


Ejemplos en Python.

S = {x² : x in {0 ... 9}}

In [52]:
S = [x**2 for x in range(10)]
S

[0, 1, 4, 9, 16, 25, 36, 49, 64, 81]

V = (1, 2, 4, 8, ..., 2¹²)

In [53]:
V = [2**i for i in range(13)]
V

[1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 2048, 4096]

M = {x | x in S and x even}

In [54]:
M = [x for x in S if (x%2==0)]
M

[0, 4, 16, 36, 64]

In [55]:
# generado por ChatGPT
squares = [i**2 for i in range(10)]
even_squares = [x for x in squares if x % 2 == 0]
print(even_squares)

[0, 4, 16, 36, 64]


Dos tipos de List Comprehension (Regular y paralela)

In [56]:
# regular list comprehension
a = [(x, y) for x in range(1, 8) for y in range(3, 6)]
a

[(1, 3),
 (1, 4),
 (1, 5),
 (2, 3),
 (2, 4),
 (2, 5),
 (3, 3),
 (3, 4),
 (3, 5),
 (4, 3),
 (4, 4),
 (4, 5),
 (5, 3),
 (5, 4),
 (5, 5),
 (6, 3),
 (6, 4),
 (6, 5),
 (7, 3),
 (7, 4),
 (7, 5)]

In [57]:
# parallel/zipped list comprehension
b = [x for x in zip(range(1, 6), range(3, 25))]
b

[(1, 3), (2, 4), (3, 5), (4, 6), (5, 7)]

### List Comprehension vs. For Loops

In [58]:
%%timeit
lista1 = []
for a in range(10000):
    b = a / 2 + 10
    lista1.append(b)

1.66 ms ± 275 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)


In [59]:
%%timeit
lista2 = [a / 2 + 10 for a in range(10000)]

1.16 ms ± 183 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)


## Referencia y copia de objetos

In [62]:
# Creación de lista y llenado usando append
lista1 = []
for a in range(10):
    b = a / 2 + 10
    lista1.append(b)

In [63]:
# Creación de lista y llenado usando "list conprehension"
lista2 = [a / 2 + 10 for a in range(10)]

In [64]:
print(lista1)
print(lista2)

[10.0, 10.5, 11.0, 11.5, 12.0, 12.5, 13.0, 13.5, 14.0, 14.5]
[10.0, 10.5, 11.0, 11.5, 12.0, 12.5, 13.0, 13.5, 14.0, 14.5]


In [65]:
print(lista1 is lista2)
print(lista1 == lista2)
1 == 2

False
True


False

In [66]:
lista3 = lista2
lista3 is lista2

True

Qué sucede si modifico un elemento de la lista2 y veo la lista3 ?

In [68]:
lista2[0] = "nuevo_elemento"

print(lista2)
print(lista3)

['nuevo_elemento', 10.5, 11.0, 11.5, 12.0, 12.5, 13.0, 13.5, 14.0, 14.5]
['nuevo_elemento', 10.5, 11.0, 11.5, 12.0, 12.5, 13.0, 13.5, 14.0, 14.5]


In [69]:
lista4 = lista1.copy()


In [70]:
lista4 is lista1

False

In [71]:
lista1[0] = "nuevo2"
lista4

[10.0, 10.5, 11.0, 11.5, 12.0, 12.5, 13.0, 13.5, 14.0, 14.5]

In [72]:
lista1

['nuevo2', 10.5, 11.0, 11.5, 12.0, 12.5, 13.0, 13.5, 14.0, 14.5]

In [73]:
lista3 = lista2

In [74]:
original_list = [1, 2, 3]
shallow_copy = original_list.copy()

# Modificamos el elemento en la posición 0 de la lista original
original_list[0] = 0

print(original_list)   # [0, 2, 3]
print(shallow_copy)    # [1, 2, 3]

[0, 2, 3]
[1, 2, 3]


In [75]:
shallow_copy[0] = 9

In [76]:
print(original_list)
print(shallow_copy)

[0, 2, 3]
[9, 2, 3]


In [84]:
import copy

old_list = [[1, 2, 3], 
            [4, 5, 6],
            [7, 8, 'a']]

# realizar una copia superficial (shallow copy)
new_list = copy.copy(old_list)

new_list[2][2] = 9

print('Old List:', old_list)
print('ID of Old List:', id(old_list))

print('New List:', new_list)
print('ID of New List:', id(new_list))

Old List: [[1, 2, 3], [4, 5, 6], [7, 8, 9]]
ID of Old List: 140389746294336
New List: [[1, 2, 3], [4, 5, 6], [7, 8, 9]]
ID of New List: 140389733020992


In [85]:
old_list = [[1, 1, 1], [2, 2, 2], [3, 3, 3]]

# realizar una copia profunda (deep copy)
new_list = copy.deepcopy(old_list)

old_list[1][0] = 'BB'

print("Old list:", old_list)
print("New list:", new_list)

Old list: [[1, 1, 1], ['BB', 2, 2], [3, 3, 3]]
New list: [[1, 1, 1], [2, 2, 2], [3, 3, 3]]


## Ejercicios 

1. Usando 'list comprehension', crear una lista con valores del 100 al 500 a intervalos de 50


In [86]:
lista_valores = [x for x in range(100, 550, 50)]
lista_valores

[100, 150, 200, 250, 300, 350, 400, 450, 500]


2. Convertir esa lista en un diccionario donde los valores sean los elementos de la lista y las claves (keys) el texto 'pago_{i}' donde i es el indice del pago, ej. pago_1 = 100. Recordar que los índices empiezan en 0.

In [87]:
dict_valores = {f"pago_{i+1}": valor for i, valor in enumerate(lista_valores)}
dict_valores

{'pago_1': 100,
 'pago_2': 150,
 'pago_3': 200,
 'pago_4': 250,
 'pago_5': 300,
 'pago_6': 350,
 'pago_7': 400,
 'pago_8': 450,
 'pago_9': 500}

3. Sacar el valor actual (VA) de la lista considerándola una lista de pagos a intervalos regulares con una tasa de interés del 100%. El valor actual es la suma de los valores individuales descontados por el interés: cada pago se descuenta dividiéndolo por (1 + tasa)^índice, por ej. 150 / ((1 + 1) ** 2). Recordar que los índices empiezan en 0.

$VA [100, 150, 200] = \dfrac{100}{(1+1)^1}+\dfrac{150}{(1+1)^2}+\dfrac{200}{(1+1)^3}$

In [88]:
VA = sum([pago / (1 + 1)**i+1 for i, pago in enumerate(lista_valores)])
VA

306.65625

## Dictionary Comprehension

Puedo crear diccionarios usando Dictionary Comprehension de la siguiente forma:

`dictionary = {key: value for vars in iterable}`

In [ ]:
dic_comp ={k: v for k, v in [("k1", 1), ("k2", 2), ("k3", 3)]}
dic_comp

{'k1': 1, 'k2': 2, 'k3': 3}

Con un for loop:

In [ ]:
cuadrado = dict()
for num in range(1, 11):
    cuadrado[num] = num*num
print(cuadrado)

{1: 1, 2: 4, 3: 9, 4: 16, 5: 25, 6: 36, 7: 49, 8: 64, 9: 81, 10: 100}


Usando Dictionary Comprehension

In [ ]:
cuadrado = {num: num*num for num in range(1, 11)}
print(cuadrado)

{1: 1, 2: 4, 3: 9, 4: 16, 5: 25, 6: 36, 7: 49, 8: 64, 9: 81, 10: 100}


### Ejercicios

1. Dado un diccionario de nombres y edades, crea un nuevo diccionario con los nombres como claves y las edades divididas en 2 como valores.

In [89]:
personas = {'Juan': 40, 'María': 30, 'Pedro': 25, 'Lucía': 20}

personas_edades_div_2 = {nombre : edad / 2 for nombre, edad in personas.items()}
personas_edades_div_2

{'Juan': 20.0, 'María': 15.0, 'Pedro': 12.5, 'Lucía': 10.0}

2. Dado un diccionario de números, crea un nuevo diccionario con las claves y valores invertidos.

In [90]:
# Diccionario original
numeros = {1: 'uno', 2: 'dos', 3: 'tres', 4: 'cuatro'}
nuevo_diccionario = {valor: clave for clave, valor in diccionario.items()}
print(nuevo_diccionario)


{1: 'k1', 2: 'k2', 3: 'k3'}


3. Filtrar un diccionario para incluir solo las entradas cuyos valores son números pares:

In [91]:
numeros = {'uno': 1, 'dos': 2, 'tres': 3, 'cuatro': 4, 'cinco': 5, 'seis': 6}

numeros_pares = {k: v for k, v in numeros.items() if v % 2 == 0}
numeros_pares

{'dos': 2, 'cuatro': 4, 'seis': 6}

4. Usando Dictionary Comprehension crear un diccionario con claves entre 1 y 10, y valores en un rango entre 100 y 500, en pasos de 20. Pista: usar `zip()`

In [95]:
diccionario = {clave: valor for clave, valor in zip(range(1, 11), range(100, 500, 20))}

print(diccionario)

{1: 100, 2: 120, 3: 140, 4: 160, 5: 180, 6: 200, 7: 220, 8: 240, 9: 260, 10: 280}


In [98]:
# otro modo de crear el diccionario usando un bucle for
diccionario = {}
for i in range(1, 11):
    diccionario[i] = 80 + i * 20

In [99]:
diccionario

{1: 100,
 2: 120,
 3: 140,
 4: 160,
 5: 180,
 6: 200,
 7: 220,
 8: 240,
 9: 260,
 10: 280}